src/QirRuntime/experimental/IonqTranslator.cpp

```diff
@@ -0,0 +1,664 @@
+            PauliId axis,
+            Qubit qubit,
+            long numerator,
+            int power) override
+        {
+            this->owner->ControlledR(numControls, controls, axis, qubit, CalculateAngle(numerator, power));
+        }
+
+        void R1Fraction(Qubit qubit, long numerator, int power) override
+        {
+            this->owner->R1(qubit, CalculateAngle(numerator, power));
+        }
+
+        void ControlledR1Fraction(long numControls, Qubit* const controls, Qubit qubit, long numerator, int power)
+            override
+        {
+            this->owner->ControlledR1(numControls, controls, qubit, CalculateAngle(numerator, power));
+        }
+
+        // TODO: implement exponents
+    };
+
+    /*==============================================================================
+        CIonqTranslator
+    ==============================================================================*/
+    class CIonqTranslator final : public ITranslator
+    {
+        // List of IonQ instructions that correspond to the commands executed by the
+        // translator.
+        vector<Instruction> instructions;
+
+        // Errors, encountered by the translator (might or might not be ignorable).
+        vector<TranslationError> errors;
+
+        // IonQ allocates all qubits upfront and doesn't reuse them, so the translator
+        // keeps track of the number of allocations to specify the size of quantum register.
+        long quantumRegisterSize = 0;
+
+      public:
+        CIonqTranslator() = default;
+        ~CIonqTranslator() = default;
+
+        void AddInstruction(Instruction&& instr)
+        {
+            if (instr.gate == Gate::alloc)
+            {
+                this->quantumRegisterSize++;
+            }
+            else
+            {
+                instructions.push_back(std::move(instr));
+            }
+        }
+
+        void AddError(TranslationError&& error)
+        {
+            this->errors.push_back(std::move(error));
+        }
+
+        void PrintRepresentation(ostringstream& os, ostringstream* err) override
+        {
+            os << "{\"qubits\":" << this->quantumRegisterSize << ",\"circuit\":[";
+            for (const Instruction& inst : this->instructions)
+            {
+                inst.Print(os);
+            }
+            os << "]}";
+
+            if (err != nullptr && !this->errors.empty())
+            {
+                *err << "{\"error_count\":" << this->errors.size() << ",\"errors\":[";
+                for (const TranslationError& error : this->errors)
+                {
+                    error.Print(*err);
+                }
+                *err << "]}";
+            }
+        }
+    };
+
+    /*==============================================================================
+        CIonqSimulator
+    ==============================================================================*/
+    class CIonqSimulator final : public CQuantumApiBase

```

Comment:
Notice, that we are hiding the concrete type of the simulator from the clients: the factory methods return pointers to single interface and to allow the client access to the other interface we'd have to either implement some kind of COM-like QI or require the clients to assemble the simulator from the parts they create themselves. I've chosen the latter per "prefer composition over inheritance" principle. The part I don't like about this approach is that one cannot actually mix-and-match translators and simulators. At least, not the ones we provide right now -- the simulators _assume_ the ITranslator given to them has particular runtime type and statically casts to it. Oh, I guess, there is an approach #3 when the client _knows_ which `IQuantumApi`s are also `ITranslator`s and does the cast on their side, but it would be even less safe than what we have right now. Also, please keep in mind that the design has to support "consuming the runtime via shared library" scenario which limits us to C-style signatures for QAPI_* functions.

I'm not 100% happy with the current design and open to reviewing/changing it, provided we can meet the requirements outlined above.